src/Lecture4.hs

```diff
@@ -131,8 +137,30 @@ errors. We will simply return an optional result here.
 🕯 HINT: Use the 'readMaybe' function from the 'Text.Read' module.
 -}
 
+trimSpaces :: String -> String
+trimSpaces = reverse . dropWhile isSpace . reverse . dropWhile isSpace
+
+splitOn :: Char -> String -> [String]
+splitOn _ [] = []
+splitOn sym str@(x:xs)
+  | x == sym = splitOn sym xs
+  | otherwise = clearedPart : splitOn sym other
+  where (part, other) = break (==sym) str
+        clearedPart = trimSpaces part
+
+parseCost :: String -> Maybe Int
+parseCost stringCost
+  | isNothing parsedValue = parsedValue
+  | fromJust parsedValue >= 0 = parsedValue
+  | otherwise = Nothing
+  where parsedValue = readMaybe stringCost
+
 parseRow :: String -> Maybe Row
-parseRow = error "TODO"
+parseRow str
+  | length clearedData == 3 = let (pName:pType:pCost:_) = clearedData in Row pName <$> readMaybe pType <*> parseCost pCost
+  | otherwise = Nothing
```

Comment:
I see that you used `Applicative`! That's very cool 🔝 

As in the previous example, you can replace `length` and `let (...) = ` pattern matching with a single pattern on the top level:

```suggestion
parseRow str = case clearedData of
    [pName, pType, pCost] -> Row pName <$> readMaybe pType <*> parseCost pCost
    _ -> Nothing
```